architecture.md

```diff
@@ -0,0 +1,685 @@
+calls directly into `kata-runtime`.
+
+## Shim
+
+A container process reaper, such as Docker's `containerd-shim` or CRI-O's `conmon`,
+is designed around the assumption that it can monitor and reap the actual container
+process. As the container process reaper runs on the host, it cannot directly
+monitor a process running within a virtual machine. At most it can see the QEMU
+process, but that is not enough. With Kata Containers, `kata-shim` acts as the
+container process that the container process reaper can monitor. Therefore
+`kata-shim` needs to handle all container I/O streams (`stdout`, `stdin` and `stderr`)
+and forward all signals the container process reaper decides to send to the container
+process.
+
+`kata-shim` has an implicit knowledge about which VM agent will handle those streams
+and signals and thus acts as an encapsulation layer between the container process
+reaper and the `kata-agent`. `kata-shim`:
+
+- Connects to `kata-proxy` on a unix domain socket. The socket url is passed from
+  `kata-runtime` to `kata-shim` when the former spawns the latter along with a
+  `containerID` and `execID`. The `containerID` and `execID` are used to identify
+  the true container process that the shim process will be shadowing or representing.
+- Forwards the standard input stream from the container process reaper into
+ `kata-proxy` using grpc `WriteStdin` gRPC API.
+- Reads the standard output/error from the container process to the
+- Forwards signals it receives from the container process reaper to `kata-proxy`
+  using `SignalProcessRequest` API.
+- Monitors terminal changes and forwards them to `kata-proxy` using grpc `TtyWinResize`
+  API.
+
+
+## Networking
+
+Containers will typically live in their own, possibly shared, networking namespace.
+At some point in a container lifecycle, container engines will set up that namespace
+to add the container to a network which is isolated from the host network, but
+which is shared between containers
+
+In order to do so, container engines will usually add one end of a `virtual ethernet
+(veth)` pair into the container networking namespace. The other end of the `veth`
+pair is added to the container network.
+
+This is a very namespace-centric approach as many hypervisors (in particular QEMU)
+cannot handle `veth` interfaces. Typically, `TAP` interfaces are created for VM
+connectivity.
+
+To overcome incompatibility between typical container engines expectations
+and virtual machines, `kata-runtime` networking transparently connects `veth`
+interfaces with `TAP` ones using MACVTAP:
+
+![Kata Containers networking](arch-images/network.png)
+
+ Kata Containers supports both
+[CNM](https://github.com/docker/libnetwork/blob/master/docs/design.md#the-container-network-model)
+and [CNI](https://github.com/containernetworking/cni) for networking management.
+
+### CNM
+
+![High-level CNM Diagram](arch-images/CNM_overall_diagram.png)
+
+__CNM lifecycle__
+
+1.  RequestPool
```

Comment:
This section is confusing. Maybe a link to https://github.com/docker/libnetwork/blob/master/docs/design.md#api would help a little.